File: finagle-core/src/main/scala/com/twitter/finagle/filter/LoggingFilter.scala
package com.twitter.finagle.filter

import com.twitter.util.{Duration, Return, Throw, Stopwatch, Future}
import com.twitter.finagle.{SimpleFilter, Service}
import com.twitter.logging.Logger

trait LogFormatter[-Req, Rep] {
  def format(request: Req, reply: Rep, replyTime: Duration): String

  def formatException(request: Req, throwable: Throwable, replyTime: Duration): String
}

/**
 * A [[com.twitter.finagle.Filter]] that logs all requests according to
 * formatter.
 */
trait LoggingFilter[Req, Rep] extends SimpleFilter[Req, Rep] {
  val log: Logger
  val formatter: LogFormatter[Req, Rep]

  def apply(request: Req, service: Service[Req, Rep]): Future[Rep] = {
    val elapsed = Stopwatch.start()
    val future = service(request)
    future respond {
      case Return(reply) =>
        log(elapsed(), request, reply)
      case Throw(throwable) =>
        logException(elapsed(), request, throwable)
    }
    future
  }

  protected def log(replyTime: Duration, request: Req, reply: Rep) {
    val line = formatter.format(request, reply, replyTime)
    log.info(line)
  }

  protected def logException(replyTime: Duration, request: Req, throwable: Throwable) {
    val line = formatter.formatException(request, throwable, replyTime)
    log.info(throwable, line)
  }

}

File: finagle-example/src/main/java/com/twitter/finagle/example/java/http/LoggingFilter.java
package com.twitter.finagle.example.java.http;

import scala.Function0;

import com.twitter.finagle.Service;
import com.twitter.finagle.SimpleFilter;
import com.twitter.finagle.http.Request;
import com.twitter.finagle.http.Response;
import com.twitter.util.Duration;
import com.twitter.util.Function;
import com.twitter.util.Future;
import com.twitter.util.Stopwatch$;

/**
 * A simple Finagle filter that logs the Request total time in milliseconds.
 */
public final class LoggingFilter extends SimpleFilter<Request, Response> {

    @Override
    public Future<Response> apply(Request req, Service<Request, Response> service) {
        final Function0<Duration> start = Stopwatch$.MODULE$.start();

        return service.apply(req).map(new Function<Response, Response>() {
            @Override
            public Response apply(Response resp) {
                Duration elapsed = start.apply();
                System.out.println("Took: " + elapsed.inMilliseconds() + "ms to complete.");

                return resp;
            }
        });
    }
}
